Supported lnls-dig/fofb-ctrl-gw#1 changes

guilhermerc opened the pull request.

guilhermerc committed:
Supported lnls-dig/fofb-ctrl-gw#1 changes [1/2]: ram writing

Also some refactoring was done on examples/src/fofb_processing.c,
aiming to increase robustness on its argument checking.

guilhermerc committed:
Supported lnls-dig/fofb-ctrl-gw#1 changes [2/2]: ram reading

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,298 @@
+  OP
+};
+
+#define NUM_OF_NECESSARY_ARGS 4
```
```suggestion
enum necessary_args_idx {
  BOARD_SLOT,
  HALCS_NUMBER,
  CHANNEL,
  OP,
  NUM_OF_NECESSARY_ARGS
};
```

Assim ele se atualiza automaticamente, se vc precisar mudar algo.

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,298 @@
-    switch (opt) {
-      /* Display Help */
+  struct optional_args opt_args = {
+    .broker_endp = strdup("ipc://"DFLT_BIND_FOLDER),
```
```suggestion
    .broker_endp = strdup("ipc://" DFLT_BIND_FOLDER),
```
Um teco mais fácil de ler

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,298 @@
-      /* Display Help */
+  struct optional_args opt_args = {
+    .broker_endp = strdup("ipc://"DFLT_BIND_FOLDER),
+    .verbose = 0
+  };
+  struct necessary_args nec_args = {
+    .board_slot = 0,
+    .halcs_number = 0,
+    .channel = 0,
+  };
```
Se você está inicializando tudo para zero, pode omitir o `.verbose`, a não ser que queira ser explícito. E pode colocar `nec_args = {0};` também.

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,298 @@
-      case 'v':
-        verbose = 1;
+      case 'b':
+        opt_args.broker_endp = strdup(optarg);
```
Como conversamos, strdup com `optarg` não é necessário.

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,298 @@
       case 'o':
-        board_number_str = strdup (optarg);
+        nec_args.board_slot = strtoul(optarg, NULL, 10);
+        nec_args.is_given[BOARD_SLOT] = true;
```
Aqui e em outros pontos você não está resetando e verificando `errno` após a conversão. `strtoul` pode retornar 0 para entradas inválidas. Acho que para argumentos é menos relevante, mas é bom lembrar.

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,298 @@
+          fp = fopen(optarg, "r");
+          if(fp == NULL) {
+            fprintf(stderr, "[client:fofb_processing]: Something wrong happened"
+              " while trying to open %s\n", optarg);
```
Dá pra imprimir `strerror(errno)` aqui para ajudar o usuário, o que você acha?

Dá pra evitar digitar tudo isso usando o especificador de formato `%m`, mas que eu me lembre não é padrão em nenhum standard ainda.

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,298 @@
+            for(int i = 0;
+              i < FOFB_PROCESSING_REGS_RAM_BANK_SIZE/sizeof(uint32_t); i++) {
+                if(fgets(coeff, 20, fp) != NULL) {
+                  coeffs.data[i] = strtoul(coeff, NULL, 10);
```
Acho que aqui é importante tratar erros de `strtoul`

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,298 @@
+
+            for(int i = 0;
+              i < FOFB_PROCESSING_REGS_RAM_BANK_SIZE/sizeof(uint32_t); i++) {
+                if(fgets(coeff, 20, fp) != NULL) {
```
```suggestion
                if(fgets(coeff, sizeof coeff, fp) != NULL) {
```

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,298 @@
+                  fprintf(stderr, "[client:fofb_processing]: Something wrong"
+                    " happened while parsing %s\n", optarg);
+
+                  ret = 1;
+                  goto err_free_mem;
```
Não importa tanto, já que você está saindo do programa quando isso acontece, mas o arquivo não está sendo fechado.

guilhermerc replied:
Conforme conversamos, o arquivo está sendo fechado.

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,298 @@
+          } else {
+            fprintf(stdout, "%d", coeffs.data[i]);
+          }
+        }
```
```suggestion
        i < FOFB_PROCESSING_REGS_RAM_BANK_SIZE/sizeof(uint32_t) - 1; i++) {
        fprintf(stdout, "%d, ", coeffs.data[i]);
      }
      fprintf(stdout, "%d", coeffs.data[i]);
```

Me parece um pouco mais limpo? Aí declara `int i` fora do for para funcionar.

ericonr replied:
Um detalhe que não sei se importa é que imprimir com  `GET_COEFF` não gera um arquivo que vc possa ler de volta depois, né?

guilhermerc replied:
Sim, de fato. Tal implementação ajudaria a scriptar um teste de escrita/leitura fazendo diff dos arquivos, né?
Vou deixar essa conversa em aberto e futuramente implemento.

guilhermerc replied:
Implementado!

ericonr commented on include/hw/wb_fofb_processing_regs.h:
```diff
@@ -13,7 +13,7 @@
 #else
 #error "Unsupported compiler?"
-#endif
+ #endif
```
Espaço extra

ericonr commented on src/client/src/halcs_client_core.c:
```diff
@@ -4228,48 +4228,61 @@ PARAM_FUNC_CLIENT_READ(rtmlamp_ohwr_pi_sp_lim_inf)
-}
+    /* Check if any error occurred */
+    ASSERT_TEST(err == HALCS_CLIENT_SUCCESS,
+        "[halcs_fofb_processing_coeff_ram_bank_read] Something wrong hapenned",
```
```suggestion
        "[halcs_fofb_processing_coeff_ram_bank_read] Something wrong happened",
```

ericonr commented on src/client/src/halcs_client_core.c:
```diff
@@ -4228,48 +4228,61 @@ PARAM_FUNC_CLIENT_READ(rtmlamp_ohwr_pi_sp_lim_inf)
-    return param_client_write (self, service, FOFB_PROCESSING_OPCODE_SET_GET_RAM_ADDR, fofb_processing_ram_addr);
-}
+    halcs_client_err_e err = halcs_func_trans_exec(self,
+        FOFB_PROCESSING_NAME_COEFF_RAM_BANK_READ, service, &input, output);
```
```suggestion
        FOFB_PROCESSING_NAME_COEFF_RAM_BANK_READ, service, &input, coeffs->data);
```

Para evitar a cópia?

ericonr commented on src/sm_io/modules/fofb_processing/sm_io_fofb_processing_exp.c:
```diff
@@ -39,37 +39,69 @@
 /************************************************************/
 /***********  Specific FOFB_PROCESSING Operations ***********/
 /************************************************************/
+static int _fofb_processing_coeff_ram_bank_read (void *owner, void *args,
```
Estilo: funções estáticas não precisam de `_` no começo como namespacing

guilhermerc replied:
Perfeito, aqui só procurei manter a consistência com os outros módulos...

ericonr commented on src/sm_io/modules/fofb_processing/sm_io_fofb_processing_exp.c:
```diff
@@ -39,37 +39,69 @@
+    if(size != FOFB_PROCESSING_REGS_RAM_BANK_SIZE) {
+        err = -FOFB_PROCESSING_ERR;
+    } else {
+        memcpy(ret, coeffs.data, FOFB_PROCESSING_REGS_RAM_BANK_SIZE);
```
Pular a cópia também é uma opção, né? Usar `ret` diretamente no `smio_thsafe_client_read_block`

guilhermerc committed:
Added minor improvements

* Fixed typos
* Improved coding choices
* Properly used 'errno' on failure checkings
* Removed whitespaces

melissa-aguiar reviewed (commented):
Code reviwed and no changes suggested

guilhermerc committed:
Add access to fixed-point position reg on fofb processing

guilhermerc committed:
Removed improper header inclusion

'hw/wb_fofb_processing_regs.h' is not installed, so it can't be found by
external programs linking with halcs.

guilhermerc committed:
Fix identation

ericonr committed:
libs/llio: flush bar4 block writes.

On bar4, when writing 512 double words (the FOFB Processing RAM banks),
we observed corruption and values that weren't changed from their
previous one. Sometimes, these writes would also lead to wishbone
timeouts (all reads returning 0xffffffff). This seems to be caused by
too many writes at once overwhelming some part of the system. Since PCIe
reads cause writes to be flushed [1] [2], we can add a read after every
write, therefore making sure the write is done before proceeding and
making it impossible to do too many writes at once.

For this to work, we had to make the bar pointers volatile, as well,
since otherwise the compiler could optimize out the read. They should
always have been volatile, but that would have required too many code
changes at once.

[1] https://www.kernel.org/doc/html/latest/PCI/pci.html#mmio-space-and-write-posting
[2] https://groups.google.com/g/comp.os.vxworks/c/34_d90EfFa0

augustofg commented on examples/src/fofb_processing.c:
```diff
@@ -1,272 +1,366 @@
-  {"ram_write_en",        required_argument,   NULL, 'w'},
-  {"ram_data_in",         required_argument,   NULL, 'd'},
-  {"ram_addr",            required_argument,   NULL, 'a'},
+#include "hw/wb_fofb_processing_regs.h" // TODO: Is this include okay to be here?
```
Still necessary?

guilhermerc replied:
Nope. Removed on e03c75135f2d685f1f5710aaab048dc2aa66e25b.
